src/lenskit/reranking/fair.py

```diff
@@ -0,0 +1,223 @@
+# This file is part of LensKit.
+# Copyright (C) 2018-2023 Boise State University.
+# Copyright (C) 2023-2025 Drexel University.
+# Licensed under the MIT license, see LICENSE.md for details.
+# SPDX-License-Identifier: MIT
+
+"""
+FA*IR top-k re-ranking.
+"""
+
+# pyright: basic
+from __future__ import annotations
+
+from collections import deque
+
+import numpy as np
+from pydantic import BaseModel, Field, PositiveFloat, PositiveInt
+from scipy.stats import binom
+from typing_extensions import override
+
+from lenskit.data import Dataset, ItemList
+from lenskit.logging import Stopwatch, get_logger
+from lenskit.pipeline import Component
+from lenskit.training import Trainable, TrainingOptions
+
+_log = get_logger(__name__)
+
+
+class FairRerankerConfig(BaseModel, extra="forbid"):
+    """
+    Configuration for :class:`FairTopNReranker`.
+    """
+
+    k: PositiveInt | None = None
+    """
+    Size of the top-k segment to rerank. If ``None``, the entire
+    candidate list is reranked (unbounded top-k).
+    """
+
+    p: PositiveFloat = Field(0.5, gt=0.0, lt=1.0)
+    """
+    The target proportion of protected items in the list (0.0 ≤ p ≤ 1.0).
+    """
+
+    alpha: PositiveFloat = Field(0.1, gt=0.0, lt=1.0)
+    """
+    Family-wise Type I error across prefixes.
+    """
+
+
+class FairReranker(Component[ItemList], Trainable):
+    """
+    FA*IR top-k re-ranking algorithm.
+
+    Re-ranks a sorted candidate list to satisfy binomial lower-bound constraints
+    for a binary protected/unprotected attribute at every prefix (1..N). It expects
+    a sorted list of items, the target proportion of protected items ``p`` and
+    significance level ``alpha``.
+
+    For details, see `Zehlike et al. (2017)`_.
+
+    .. _Zehlike et al. (2017): https://doi.org/10.1145/3132847.3132938
+
+    Stability:
+        Caller
+
+    Args:
+        k:
+            Size of the top-k subset to rerank. If ``None``, the entire
+            candidate list is reranked.
+        p:
+            The target proportion of protected items in the list (0.0 ≤ p ≤ 1.0).
+            The default value is 0.5.
+        alpha:
+            The statistical significance level for the fairness test.
+            The default value is 0.1.
+
```

Comment:
We don't need to duplicate this documentation here, the configuration class is sufficient.